File: packages/nitro-renderer/src/room/object/visualization/data/SizeData.ts
import type { IAssetColor, IAssetVisualizationDirection, IAssetVisualizationLayer } from '@nitrodevco/nitro-api';
import type { BLEND_MODES } from 'pixi.js';

import { ColorData } from './ColorData';
import { DirectionData } from './DirectionData';
import { LayerData } from './LayerData';

export class SizeData {
    public static MAX_LAYERS: number = 26;

    private _layerCount: number;
    private _angle: number;

    private _defaultDirection: DirectionData;
    private _directions: Map<number, DirectionData> = new Map();
    private _colors: Map<number, ColorData> = new Map();
    private _lastDirectionData: DirectionData | undefined = undefined;
    private _lastDirection: number = -1;

    constructor(layerCount: number, angle: number) {
        this._layerCount = layerCount < 0 ? 0 : layerCount > SizeData.MAX_LAYERS ? SizeData.MAX_LAYERS : layerCount;
        this._angle = angle < 1 ? 1 : angle > 360 ? 360 : angle;

        this._defaultDirection = new DirectionData(this._layerCount);
    }

    public dispose(): void {
        if (this._defaultDirection) this._defaultDirection.dispose();

        for (const direction of this._directions.values()) direction?.dispose();

        for (const color of this._colors.values()) color?.dispose();

        this._defaultDirection = undefined!;
        this._lastDirectionData = undefined;
        this._lastDirection = -1;

        this._directions.clear();
        this._colors.clear();
    }

    public processLayers(layers: IAssetVisualizationLayer[]): boolean {
        return this.setDirectionLayers(this._defaultDirection, layers);
    }

    public processDirections(directions: IAssetVisualizationDirection[]): boolean {
        if (!directions) return false;

        for (const direction of directions) {
            if (this._directions.get(direction.id)) return false;

            const directionData = new DirectionData(this._layerCount);

            directionData.setFromDirection(this._defaultDirection);

            if (direction.layers !== undefined) this.setDirectionLayers(directionData, direction.layers);

            this._directions.set(direction.id, directionData);

            this._lastDirectionData = undefined;
            this._lastDirection = -1;
        }

        return true;
    }

    public processColors(colors: IAssetColor[]): boolean {
        if (!colors) return false;

        for (const color of colors) {
            if (this._colors.get(color.id)) return false;

            const colorData = new ColorData(this._layerCount);

            if (color.layers) {
                for (const layer of color.layers) {
                    if (layer?.color) colorData.setColorLayer(layer.id, layer.color);
                }
            }

            this._colors.set(color.id, colorData);
        }

        return true;
    }

    private setDirectionLayers(
        directionData: DirectionData,
        layers: IAssetVisualizationLayer[],
    ): boolean {
        if (!directionData || !layers) return false;

        for (const layer of layers) {
            if (layer.id < 0 || layer.id >= this._layerCount) return false;

            if (layer.ink !== undefined) directionData.setLayerBlendMode(layer.id, layer.ink?.toLowerCase() as BLEND_MODES);

            if (layer.tag !== undefined) directionData.setLayerTag(layer.id, layer.tag);

            if (layer.alpha !== undefined) directionData.setLayerAlpha(layer.id, layer.alpha);

            if (layer.ignoreMouse !== undefined) directionData.setLayerIgnoreMouse(layer.id, layer.ignoreMouse);

            if (layer.x !== undefined) directionData.setLayerXOffset(layer.id, layer.x);

            if (layer.y !== undefined) directionData.setLayerYOffset(layer.id, layer.y);

            if (layer.z !== undefined) directionData.setLayerZOffset(layer.id, layer.z / -1000);
        }

        return true;
    }

    public getValidDirection(direction: number): number {
        const existing = this._directions.get(direction);

        if (existing) return direction;

        direction = ((direction % 360) + 360) % 360;

        let currentAngle = -1;
        let validDirection = -1;

        for (const key of this._directions.keys()) {
            let angle = (key * this._angle - direction + 360) % 360;

            if (angle > 180) angle = 360 - angle;

            if (angle < currentAngle || currentAngle < 0) {
                currentAngle = angle;
                validDirection = key;
            }
        }

        if (validDirection >= 0) return Math.trunc(validDirection);

        return 0;
    }

    public getDirectionData(direction: number): DirectionData {
        if (direction === this._lastDirection && this._lastDirectionData) return this._lastDirectionData;

        const directionData = this._directions.get(direction) ?? this._defaultDirection;

        this._lastDirection = direction;
        this._lastDirectionData = directionData;

        return this._lastDirectionData;
    }

    public getLayerTag(direction: number, layerId: number): string {
        return this.getDirectionData(direction)?.getLayerTag(layerId) ?? LayerData.DEFAULT_TAG;
    }

    public getLayerBlendMode(direction: number, layerId: number): BLEND_MODES {
        return this.getDirectionData(direction)?.getLayerBlendMode(layerId) ?? LayerData.DEFAULT_BLEND_MODE;
    }

    public getLayerAlpha(direction: number, layerId: number): number {
        return this.getDirectionData(direction)?.getLayerAlpha(layerId) ?? LayerData.DEFAULT_ALPHA;
    }

    public getLayerIgnoreMouse(direction: number, layerId: number): boolean {
        return this.getDirectionData(direction)?.getLayerIgnoreMouse(layerId) ?? LayerData.DEFAULT_IGNORE_MOUSE;
    }

    public getLayerXOffset(direction: number, layerId: number): number {
        return this.getDirectionData(direction)?.getLayerXOffset(layerId) ?? LayerData.DEFAULT_XOFFSET;
    }

    public getLayerYOffset(direction: number, layerId: number): number {
        return this.getDirectionData(direction)?.getLayerYOffset(layerId) ?? LayerData.DEFAULT_YOFFSET;
    }

    public getLayerZOffset(direction: number, layerId: number): number {
        return this.getDirectionData(direction)?.getLayerZOffset(layerId) ?? LayerData.DEFAULT_ZOFFSET;
    }

    public getLayerColor(layerId: number, colorId: number): number {
        const existing = this._colors.get(colorId);

        if (!existing) return ColorData.DEFAULT_COLOR;

        return existing.getLayerColor(layerId);
    }

    public get layerCount(): number {
        return this._layerCount;
    }
}

File: packages/nitro-swift/Sources/NitroRoom/Furniture/SizeData.swift
import Foundation

/// Swift port of `SizeData` (packages/nitro-renderer/src/room/object/visualization/data/SizeData.ts).
/// One instance per declared `size` (32/64/1) within a furniture's `IAssetVisualizationData[]`.
/// Not `final` - `AnimationSizeData` subclasses it, matching the original's `AnimationSizeData extends SizeData`.
public class SizeData {
    public static let maxLayers = 26

    private let _layerCount: Int
    private let angle: Double

    private let defaultDirection: DirectionData
    private var directions: [Int: DirectionData] = [:]
    private var colors: [Int: ColorData] = [:]

    private var lastDirectionData: DirectionData?
    private var lastDirection: Int = -1

    public init(layerCount: Int, angle: Double) {
        _layerCount = min(max(layerCount, 0), SizeData.maxLayers)
        self.angle = min(max(angle, 1), 360)
        defaultDirection = DirectionData(layerCount: _layerCount)
    }

    @discardableResult
    public func processLayers(_ layers: [AssetVisualizationLayer]) -> Bool {
        setDirectionLayers(defaultDirection, layers)
    }

    @discardableResult
    public func processDirections(_ directionList: [AssetVisualizationDirection]) -> Bool {
        for direction in directionList {
            if directions[direction.id] != nil { return false }

            let directionData = DirectionData(layerCount: _layerCount)

            directionData.setFrom(defaultDirection)

            if let layers = direction.layers { setDirectionLayers(directionData, layers) }

            directions[direction.id] = directionData
            lastDirectionData = nil
            lastDirection = -1
        }

        return true
    }

    @discardableResult
    public func processColors(_ colorList: [AssetColor]) -> Bool {
        for color in colorList {
            if colors[color.id] != nil { return false }

            let colorData = ColorData(layerCount: _layerCount)

            for layer in color.layers ?? [] {
                if let value = layer.color { colorData.setColorLayer(layer.id, value) }
            }

            colors[color.id] = colorData
        }

        return true
    }

    @discardableResult
    private func setDirectionLayers(_ directionData: DirectionData, _ layers: [AssetVisualizationLayer]) -> Bool {
        for layer in layers {
            guard layer.id >= 0, layer.id < _layerCount else { return false }

            if let ink = layer.ink { directionData.setLayerBlendMode(layer.id, ink.lowercased()) }
            if let tag = layer.tag { directionData.setLayerTag(layer.id, tag) }
            if let alpha = layer.alpha { directionData.setLayerAlpha(layer.id, alpha) }
            if let ignoreMouse = layer.ignoreMouse { directionData.setLayerIgnoreMouse(layer.id, ignoreMouse) }
            if let x = layer.x { directionData.setLayerXOffset(layer.id, x) }
            if let y = layer.y { directionData.setLayerYOffset(layer.id, y) }
            if let z = layer.z { directionData.setLayerZOffset(layer.id, z / -1000) }
        }

        return true
    }

    public func getValidDirection(_ requested: Int) -> Int {
        if directions[requested] != nil { return requested }

        let direction = Double(((requested % 360) + 360) % 360)

        var currentAngle: Double = -1
        var validDirection = -1

        for key in directions.keys {
            var a = (Double(key) * angle - direction + 360).truncatingRemainder(dividingBy: 360)

            if a > 180 { a = 360 - a }

            if a < currentAngle || currentAngle < 0 {
                currentAngle = a
                validDirection = key
            }
        }

        return validDirection >= 0 ? validDirection : 0
    }

    public func getDirectionData(_ direction: Int) -> DirectionData {
        if direction == lastDirection, let cached = lastDirectionData { return cached }

        let data = directions[direction] ?? defaultDirection

        lastDirection = direction
        lastDirectionData = data

        return data
    }

    public func getLayerTag(_ direction: Int, _ layerId: Int) -> String { getDirectionData(direction).getLayerTag(layerId) }
    public func getLayerBlendMode(_ direction: Int, _ layerId: Int) -> String { getDirectionData(direction).getLayerBlendMode(layerId) }
    public func getLayerAlpha(_ direction: Int, _ layerId: Int) -> Double { getDirectionData(direction).getLayerAlpha(layerId) }
    public func getLayerIgnoreMouse(_ direction: Int, _ layerId: Int) -> Bool { getDirectionData(direction).getLayerIgnoreMouse(layerId) }
    public func getLayerXOffset(_ direction: Int, _ layerId: Int) -> Double { getDirectionData(direction).getLayerXOffset(layerId) }
    public func getLayerYOffset(_ direction: Int, _ layerId: Int) -> Double { getDirectionData(direction).getLayerYOffset(layerId) }
    public func getLayerZOffset(_ direction: Int, _ layerId: Int) -> Double { getDirectionData(direction).getLayerZOffset(layerId) }

    public func getLayerColor(_ layerId: Int, _ colorId: Int) -> UInt32 {
        colors[colorId]?.getLayerColor(layerId) ?? ColorData.defaultColor
    }

    public var layerCount: Int { _layerCount }
}
